### ISSUE TITLE
Ability to have one global cache file

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#4241]
I want to take advantage of the `cache` and `cacheFile` option, but I don't want to litter every single project with a `.eslintcache` file and having to also .gitignore it. I would much rather have one global cache file in my home folder that is responsible for caching all ESLint runs. This could easily be achieved by ESLint modifying the cache file instead of overwriting. I could then just point the `cacheFile` option to a file in my home folder. This is how Babel does caching fwiw.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by kevva** [comment]
:+1:

**Comment by nzakas** [comment]
@royriojas what do you think?

**Comment by royriojas** [comment]
@nzakas, I agree we can have a global location for the cache.

@sindresorhus  I'm biased to actually have a cache folder instead of a cache file, since that will make very easy to implement this behavior. Something like `~/.eslintcache/` in the home folder of the user.

We can add a --global-cache option that will use that directory if possible.

What do you guys think?

**Comment by sindresorhus** [comment]
How would it work with a cache folder? One file in the folder for each linted path? That would actually be beneficial, as one cache file would grow very large over time.

**Comment by royriojas** [comment]
@sindresorhus yes, just like that

**Comment by nzakas** [comment]
Isn't `.eslintcache` already a folder?

I'd rather not have a new option, couldn't we just use `--cache-file` to point to a location where the cache can be stored? If so, then having multiple projects point to the same file should create a de facto "global" cache if, as @sindresorhus suggests, we always add to it instead of overwriting. (?)

**Comment by royriojas** [comment]
@nzakas no it wasn't 

the problem with that is that the file might grow really big overtime. I suggest having a cache file per path where eslint is executed.

example:

```bash
# when run from this location /Users/projects/project-name
eslint --cache-file='~/.eslintcache/' **/*.js

# this will create a file for that given location inside 
# the provided folder (has to end with a / to be consider 
# a directory). The file name will be the hash of the location so 
# if executed again on the same set of files will not create a new 
# file but reuse the existing one.
```

**Comment by royriojas** [comment]
The way the cache works it will drop any file entry that is not in the set of files, this is to avoid having cached files that were removed. That will make harder to reuse a single file, since from two directories the files are different, so when executed from directoryA the files will be cached, but when executed from directoryB the files will be different, so defeating the cache. 

A directory with cache files that correspond to an execution path will be a safer bet. I guess.

What do you think?

**Comment by vdemedes** [comment]
+1! Would be a good improvement indeed.

**Comment by nzakas** [comment]
My big concern here is introducing a breaking change - changing change line arguments is particularly hairy because they are often set deep in CI systems. While I'm not opposed to the proposal, I'd like a way to be backwards compatible.

**Comment by royriojas** [comment]
@nzakas true, that's why I believe reusing the `cache-file` option to point to a directory instead to a file could be a solution, meaning if the path ends in `/` then we assume is a directory and use it to store the cache file inside. I guess that would still be backward compatible.

**Comment by nzakas** [comment]
We should probably use `fs` to determine if it's actually a directory or not, fuzzy testing for slashes is a poor heuristic.

**Comment by royriojas** [comment]
@nzakas, I do use the `fs` to determine it, but the problem is that after doing this

```javascript
path.resolve('./some/path/.cache/') 
// this will return ./some/path/.cache
```

So this is only an issue in case the filed didn't exist. Because in that case the user might expect a folder to be created but since the `path.resolve` removed the trailing `path.sep` then we end creating a file.

So this is more like a corner case, but the first check does use the `fs.lstats` to determine if it is a directory.

**Comment by nzakas** [comment]
Gotcha

**Comment by jonathanolson** [comment]
Yes, not fully overwriting the cache file would be great. If we alternate between "lint just a few files" and "lint all the files", the cache will get mostly cleared when we only lint a few of the files.

The existence of a default global cache location would just be gravy.

**Comment by royriojas** [comment]
@jonathanolson currently as part of the work on this issue I have added the option to create the cache based on the current directory, but it seems that what you actually need is to specify different cache files for different set of files. Something like:

```bash
# all files
eslint 'src/**/*.js' --cache --cache-location='./cache-file-all'

# only the ones in lib
eslint 'src/lib/**/*.js' --cache --cache-location='./cache-file-lib'
```

by default files that are "not seen between executions" are just removed from the cache, this is to prevent the cache from keeping metadata from removed files.

**Comment by sindresorhus** [comment]
@royriojas Outside the scope of this issue, but I think it would be useful to implement a [LRU cache](https://github.com/isaacs/node-lru-cache) instead of evicting unused files right away. That would solve the above use-case. Happy to open a new issue if there's interest?

**Comment by royriojas** [comment]
@sindresorhus actually I guess you're right. There is not a real reason to evict the not seen files immediately. an LRU with a given maxAge might help in this case.

The initial implementation was, by design, super simple. but yeah it might be something useful for cases where different set of files are linted from the same directory.
